File: example/src/main/java/ru/virgil/example/security/SecurityUserDetailsService.java
package ru.virgil.example.security;

import lombok.Getter;
import lombok.RequiredArgsConstructor;
import org.springframework.context.annotation.Primary;
import org.springframework.security.authentication.AuthenticationServiceException;
import org.springframework.security.core.Authentication;
import org.springframework.stereotype.Service;
import ru.virgil.example.user.UserDetails;
import ru.virgil.example.user.UserDetailsService;
import ru.virgil.security.entity.SecurityUser;
import ru.virgil.security.firebase.FirebaseAuthenticationToken;

import java.util.Set;

@Primary
@Service
@RequiredArgsConstructor
public class SecurityUserDetailsService implements ru.virgil.security.service.SecurityUserService {

    @Getter
    private final SecurityUserDetailsRepository repository;
    private final UserDetailsService userDetailsService;
    private final OnUserRegistrationMockService onUserRegistrationMockService;

    @Override
    public SecurityUser registerOrLogin(Authentication authentication) {
        if (!(authentication instanceof FirebaseAuthenticationToken firebaseAuthenticationToken)) {
            throw new AuthenticationServiceException("Wrong %s token type in %s".formatted(
                    authentication.getClass().getSimpleName(), this.getClass().getSimpleName()));
        }
        return repository.findByFirebaseUserId(firebaseAuthenticationToken.getPrincipal().toString())
                .orElseGet(() -> register(firebaseAuthenticationToken));
    }

    public SecurityUserDetails register(FirebaseAuthenticationToken firebaseAuthenticationToken) {
        SecurityUserDetails securityUserDetails = new SecurityUserDetails();
        securityUserDetails.setFirebaseUserId(firebaseAuthenticationToken.getPrincipal().toString());
        securityUserDetails.setAuthorities(Set.of(UserAuthority.USER));
        securityUserDetails = repository.save(securityUserDetails);
        UserDetails userDetails = new UserDetails();
        userDetails.setSecurityUserDetails(securityUserDetails);
        userDetails = userDetailsService.getRepository().save(userDetails);
        userDetails = onUserRegistrationMockService.mock(userDetails);
        userDetails = userDetailsService.getRepository().save(userDetails);
        return securityUserDetails;
    }

}

File: example/src/main/java/ru/virgil/example/security/SecurityUserDetailsService.kt
package ru.virgil.example.security

import org.springframework.context.annotation.Primary
import org.springframework.security.authentication.AuthenticationServiceException
import org.springframework.security.core.Authentication
import org.springframework.stereotype.Service
import ru.virgil.example.system.mock.Mocker
import ru.virgil.example.user.UserDetails
import ru.virgil.example.user.UserDetailsRepository
import ru.virgil.example.user.UserDetailsService
import ru.virgil.security.entity.SecurityUser
import ru.virgil.security.firebase.FirebaseAuthenticationToken
import ru.virgil.security.service.SecurityUserService
import ru.virgil.utils.Faker

@Primary
@Service
class SecurityUserDetailsService(
    private val repository: SecurityUserDetailsRepository,
    private val userDetailsService: UserDetailsService,
    private val userDetailsRepository: UserDetailsRepository,
    private val faker: Faker,
    private val mocker: Mocker,
) : SecurityUserService {

    override fun registerOrLogin(authentication: Authentication): SecurityUser {
        if (authentication !is FirebaseAuthenticationToken) {
            throw AuthenticationServiceException(
                "Wrong %s token type in %s".format(
                    authentication.javaClass.simpleName, this.javaClass.simpleName
                )
            )
        }
        return repository.findByFirebaseUserId(authentication.principal.toString())
            .orElseGet { register(authentication) }
    }

    fun register(firebaseAuthenticationToken: FirebaseAuthenticationToken): SecurityUserDetails {
        var securityUserDetails = SecurityUserDetails(
            userAuthorities = mutableSetOf(UserAuthority.ROLE_USER),
            firebaseUserId = firebaseAuthenticationToken.principal.toString()
        )
        securityUserDetails = repository.save(securityUserDetails)
        var userDetails = UserDetails(securityUserDetails)
        userDetails.name = faker.animal().name()
        userDetails = userDetailsRepository.save(userDetails)
        userDetails = mock(userDetails)
        return securityUserDetails
    }

    private fun mock(userDetails: UserDetails): UserDetails {
        userDetailsService.mockingTimeUserDetails = userDetails
        mocker.mock()
        return userDetails
    }
}
